### PR TITLE
Fix behavior for Customer REST API to match expectation

### LINKED ISSUE #14054: Updating customers through rest/V1/customers/:id fails since 2.2.2
### Preconditions
1. We first tested this on version 2.2.1 -- updating customers through the API works.
2. We then upgraded to version 2.2.2 -- updating customers through the API fails.

The error persists in 2.2.3.

### Steps to reproduce
1. Create a customer through the API.
2. Update a customer through the API.

We use Python to interface with Magento.

Our update call is:

```
requests.put('{}/rest/V1/customers/{}'.format(
                    api.endpoint, customer_id), headers=api.headers,
                    json=customer)
```

The customer-value passed onto the json argument is:

```
{'customer': {'store_id': 1, 'website_id': 1, 'group_id': 1, 'firstname': 'Anonymous', 'lastname': 'Anonymous', 'email': 'valid@email.org', 'taxvat': None, 'addresses': [{'firstname': 'Anonymous', 'lastname': 'Anonymous', 'telephone': '0123456789', 'company': 'Anonymous Inc.', 'street': ['Where the streets have no name 1'], 'postcode': '1000AA', 'city': 'Amsterdam', 'country_id': 'NL', 'default_billing': False, 'default_shipping': False}, {'firstname': 'Anonymous', 'lastname': 'Anonymous', 'telephone': '0123456789', 'company': 'Anonymous BV', 'street': ['Where the streets have no name 1'], 'postcode': '1000AA', 'city': 'Amsterdam', 'country_id': 'NL', 'default_billing': False, 'default_shipping': False}], 'custom_attributes': [{'attribute_code': 'external_reference', 'value': 10011}]}}
```

I have made some adjustments to the above to filter out sensitive data. Hopefully I didn't mess up the json data itself but our import worked on Magento 2.2.1.

### Expected result

JSON including the updated customer.

### Actual result

```
{'message': 'A customer with the same email already exists in an associated website.', 'trace': etc.
```

### LINKED ISSUE #21237: Update Customer data with REST API
### Preconditions (*)

1. 2.3.0 (have not tested previous versions)

### Steps to reproduce (*)

1. When using the API to update a customer

     curl -X PUT "https://example.com/index.php/rest/all/V1/customers/8" -H "accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer TOKEN" -d "{ \"customer\": { \"middlename\": \"string\" }}"

### Expected result (*)

1. The customer should be updated for those fields just as when using the API to update a product  a PUT request like this 

    curl -X PUT "https://example.com/index.php/rest/all/V1/products/24-MB01" -H "accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer TOKEN" -d "{ \"product\": { \"visibility\": 3 }}"

updates the visibility for the product with that SKU (as one would expect) however when i try to make a similar update to a customer with a PUT request i get:
### Actual result (*)

1. I get a 400 error with `"message": "The customer email is missing. Enter and try again."` the API to update only works if i add all of the required data like this:

    curl -X PUT "https://example.com/index.php/rest/all/V1/customers/8" -H "accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer TOKEN" -d "{ \"customer\": { \"id\": \"8\", \"firstname\": \"Jon\", \"lastname\": \"Smith\", \"email\": \"Jjsmith@gmail.com\", \"middlename\": \"jac\", \"store_id\": 1, \"website_id\": 1 }}"

if I leave any of these values out i get an error with the exception of `id` which if left out creates a new record or throws an error if the email already exists (despite the `id` being specified at the end of the URL) also when `id` is passed in with the data that IDs record is updated even if it is not the `id` at the end of the URL

This does not seem to be the way the API should operate any help or explanation of whats going on would be appreciated.

### PR DESCRIPTION
**Pull request by judaschwartz**
### Description (*)

When the rest API is used to update an existing customer you must enter the firstname, lastname, email and website_id for it to work 
### Fixed Issues (if relevant)

1. Fixes magento/magento2#21237 : Update Customer data with REST API
2. Fixes  https://github.com/magento/magento2/issues/14054 : Updating customers through rest/V1/customers/:id fails since 2.2.2

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [*] Pull request has a meaningful description of its purpose
 - [*] All commits are accompanied by meaningful commit messages
 - [*] All new or changed code is covered with unit/integration tests (if applicable)
 - [*] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=21236) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @judaschwartz. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by torhoehn** [comment]
@judaschwartz Thanks for your creating your first PR! Please read the [Contribution Requirements](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#requirements) and update your PR (E.g. manual testing scenarios). Maybe also think about creating a issue to describe your expected behaviour and the actual result, to see if it's confirmed.

**Comment by judaschwartz** [comment]
> @judaschwartz Thanks for your creating your first PR! Please read the [Contribution Requirements](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#requirements) and update your PR (E.g. manual testing scenarios). Maybe also think about creating a issue to describe your expected behaviour and the actual result, to see if it's confirmed.

@torhoehn see here for the issue:
https://magento.stackexchange.com/questions/258640/update-customer-data-with-rest-api
 as far as other things like "manual testing scenarios" as you have noted this is my first PR and i am unsure what else it needs prehaps you can tell me.

Thanks,
Yehuda

**Comment by torhoehn** [comment]
@judaschwartz Because there's no way to confirm a issue or not I would like you to create an issue for that here: https://github.com/magento/magento2/issues/new/choose

**Comment by judaschwartz** [comment]
@torhoehn 
i added an issue
https://github.com/magento/magento2/issues/21237

**Comment by torhoehn** [comment]
@judaschwartz Thank you!

**Comment by judaschwartz** [comment]
Issue has been accepted

**Comment by torhoehn** [comment]
@judaschwartz For your contribution some unit tests failed. Can you fix this?

**Comment by judaschwartz** [comment]
@torhoehn ok i fixed the unit test sorry about that

**Comment by judaschwartz** [comment]
@torhoehn it passed all the unit tests

**Comment by judaschwartz** [comment]
@torhoehn any update on the status?

**Comment by treestonemedia** [comment]
bump

**Comment by judaschwartz** [comment]
@torhoehn  any update on the status of this?

File: app/code/Magento/Customer/etc/webapi.xml
Comment (sidolov) [inline app/code/Magento/Customer/etc/webapi.xml:131]: Unfortunately, we can't change webapi endpoint according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

**Review by sidolov — CHANGES_REQUESTED** [review]

File: app/code/Magento/Customer/etc/webapi.xml
Comment (judaschwartz) [reply app/code/Magento/Customer/etc/webapi.xml:131]: @sidolov  I don't see how this breaks backward compatibility, I am not removing the PUT method I am just adding a POST one

Comment (torhoehn) [reply app/code/Magento/Customer/etc/webapi.xml:131]: @judaschwartz But you changed `<route url="/V1/customers/:customerId" method="PUT">` to `<route url="/V1/customers/:id" method="PUT">`. I think is what @sidolov means.

Comment (judaschwartz) [reply app/code/Magento/Customer/etc/webapi.xml:131]: @torhoehn that change is actually a revert to how it was in 2.0 before someone changed it and broke this functionality https://github.com/magento/magento2/blob/2.0/app/code/Magento/Customer/etc/webapi.xml#L125

**Comment by taqvi** [comment]
I am having same issue on Magento CE 2.3

**Comment by marvinhinz** [comment]
What needs to be done to merge this for the next release?

**Comment by judaschwartz** [comment]
I dont know i think @sidlov needs to unblock merging

**Comment by ihor-sviziev** [comment]
@sidolov  this pull request got stuck. Could you update status of it?

**Comment by mmihalev** [comment]
I can't believe that more than one year nobody can accept and merge this PR. This is a serious API flaw and I think that it should be finally fixed!

**Comment by gotanasro3** [comment]
please fix it

File: app/code/Magento/Catalog/Api/ProductRepositoryInterface.php
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Api/ProductRepositoryInterface.php]: I prefer to revert this change, as this interface used not only for web api, but also as API for external modules.

File: app/code/Magento/Customer/Api/CustomerRepositoryInterface.php
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Api/CustomerRepositoryInterface.php]: I prefer to revert this change, as this interface used not only for web api, but also as API for external modules.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @judaschwartz,
Sorry for really long delay since you created this PR. Unfortunately this PR was in status "on hold" without any comments and it wasn't clear why. I reviewed it again - seems like good solution.

Please review and fix all types of tests (functional test failures looks not related).

File: app/code/Magento/Catalog/Api/ProductRepositoryInterface.php
Comment (judaschwartz) [reply app/code/Magento/Catalog/Api/ProductRepositoryInterface.php]: Reverted

**Comment by judaschwartz** [comment]
> Hi @judaschwartz,
> Sorry for really long delay since you created this PR. Unfortunately this PR was in status "on hold" without any comments and it wasn't clear why. I reviewed it again - seems like good solution.
> 
> Please review and fix all types of tests (functional test failures looks not related).

Yes the long delay is ridiculous, I no longer work with m2 and will not be going back to make any additions to this pr to satisfy any new test requirements that have been added in the meantime  ¯\_(ツ)_/¯ 
I did however revert the comments as you requested

**Comment by ihor-sviziev** [comment]
Hi @judaschwartz 
I reviewed test failures and unfortunately looks your changes breaks some functionality in some cases, so we can’t accept PR in current state.
I see that you don’t work with magento 2 anymore and will not be able to update this PR to get it merged, so in this case I think we can only close this PR.

I’ve linked this PR to the issue, so everyone will see it in the issue and could create new PR based on your solution and fix the issues that were found by automated tests.

We’re really sorry for huge delay in processing your PR, we’ll try to do our best in order to improve situation.

**Comment by treestonemedia** [comment]
@ihor-sviziev 
way to go!
let a PR sit for two years, then wake up one day and agree that it's a needed fix to your codebase but then close it...

**Comment by mmihalev** [comment]
Yeah, let's just close it. That would "fix" the issue...

**Pull request closed without merging**
